docs/usage/types/custom.md

```diff
@@ -2,6 +2,283 @@
+This new type can be as simple as a name or have complex validation logic attached to it:
+
+```py
+from typing import List, TypeVar
+
+from annotated_types import Gt
+from typing_extensions import Annotated, TypeAliasType
+
+from pydantic import BaseModel
+
+T = TypeVar('T')  # or a `bound=SupportGt`
+
+ImplicitAliasPositiveIntList = List[Annotated[int, Gt(0)]]
+
+
+class Model1(BaseModel):
+    x: ImplicitAliasPositiveIntList
+    y: ImplicitAliasPositiveIntList
+
+
+print(Model1.model_json_schema())
+"""
+{
+    'properties': {
+        'x': {
+            'items': {'exclusiveMinimum': 0, 'type': 'integer'},
+            'title': 'X',
+            'type': 'array',
+        },
+        'y': {
+            'items': {'exclusiveMinimum': 0, 'type': 'integer'},
+            'title': 'Y',
+            'type': 'array',
+        },
+    },
+    'required': ['x', 'y'],
+    'title': 'Model1',
+    'type': 'object',
+}
+"""
+
+PositiveIntList = TypeAliasType('PositiveIntList', List[Annotated[int, Gt(0)]])
+
+
+class Model2(BaseModel):
+    x: PositiveIntList
+    y: PositiveIntList
+
+
+print(Model2.model_json_schema())
+"""
+{
+    '$defs': {
+        'PositiveIntList': {
+            'items': {'exclusiveMinimum': 0, 'type': 'integer'},
+            'type': 'array',
+        }
+    },
+    'properties': {
+        'x': {'$ref': '#/$defs/PositiveIntList'},
+        'y': {'$ref': '#/$defs/PositiveIntList'},
+    },
+    'required': ['x', 'y'],
+    'title': 'Model2',
+    'type': 'object',
+}
+"""
+```
+
+#### Named recursive types
+
+You can also use `TypeAliasType` to create recursive types:
+
+```py
+from typing import Any
+
+from pydantic_core import PydanticCustomError
+from typing_extensions import Annotated, TypeAliasType
+
+from pydantic import (
+    TypeAdapter,
+    ValidationError,
+    ValidationInfo,
+    ValidatorFunctionWrapHandler,
+    WrapValidator,
+)
+
+
+def json_custom_error_validator(
+    value: Any, handler: ValidatorFunctionWrapHandler, _info: ValidationInfo
+) -> Any:
+    """Simplify the error message to avoid a gross error stemming
+    from exhaustive checking of all union options.
+    """
+    try:
+        return handler(value)
+    except ValidationError:
+        raise PydanticCustomError(
+            'invalid_json',
+            'Input is not valid json',
+        )
+
+
+Json = TypeAliasType(
+    'Json',
+    Annotated[
+        dict[str, 'Json'] | list['Json'] | str | int | float | bool | None,
+        WrapValidator(json_custom_error_validator),
+    ],
+)
+
+
+ta = TypeAdapter(Json)
+
+v = ta.validate_python({'x': [1], 'y': {'z': True}})
+assert v == {'x': [1], 'y': {'z': True}}
+
+try:
+    ta.validate_python({'x': object()})
+except ValidationError as exc:
+    print(exc)
+    """
+    1 validation error for function-wrap[json_custom_error_validator()]
+      Input is not valid json [type=invalid_json, input_value={'x': <object object at 0x0123456789ab>}, input_type=dict]
+    """
+```
```

Comment:
@JelleZijlstra you might find this section interesting, also thank you for your work on this!